### ISSUE TITLE
Support authentication of polymorphic Principal classes

### ISSUE THREAD

**Issue by evnm** [dropwizard/dropwizard#1318]
A web service that I'm implementing requires the notion of different classes of API keys for authenticating requests to different endpoints (e.g. read vs write keys). I thought that the natural way to implement this would be to form a simple class hierarchy of `java.security.Principal`s:

```
public abstract class ApiKey implements Principal { ... }
public class ReadApiKey extends ApiKey { ... }
public class WriteApiKey extends ApiKey { ... }
```

And then use the `@Auth` annotation to inject these keys in endpoints accordingly:

```
@POST public Response writeSomething(@Auth WriteApiKey apiKey, ...) { ... }
@GET public Response readSomething(@Auth ReadApiKey apiKey, ...) { ... }
```

However, Dropwizard is [currently](https://github.com/dropwizard/dropwizard/blob/23f353a896c9e5ad5e963fece3f82d227b7c0f5f/dropwizard-auth/src/main/java/io/dropwizard/auth/AuthValueFactoryProvider.java#L55) unable to achieve this because `io.dw.auth.AuthValueFactoryProvider` enforces that only a single Principal implementation can be used per application.

I got around this deficiency by writing a custom `AbstractValueFactoryProvider` subclass that operates on a `Set<? extends Principal>` rather than a single Principal. This seems like a useful thing to have in dropwizard-auth proper, so I was wondering if others would entertain the idea of introducing it. To make sure this is feasible, I [moved the class into dropwizard-auth as a superclass of `io.dw.auth.AuthValueFactoryProvider`](https://github.com/evnm/dropwizard/commit/f1194a02271e1353945fedaa1c7a07f9a4265e53), since it's technically a generalization thereof.

Please let me know if this is a) ill-advised or b) worthy of a PR.

**Comment by PixelBumper** [comment]
+1 I hit the wall because of this too.

**Comment by jplock** [comment]
Fixed by #1392
